### PR TITLE
feat: support `bn256_pairing` precompiled contract by Godwoken syscall

### PR DESCRIPTION
**Pull request by Flouse**
Previously, `bn256_pairing` precompiled contract was not supported due to too high cycle cost.

This PR implements `bn256_pairing` precompiled contract by introducing new Godwoken syscall:
```markdown
sys_bn_pairing
```

## Dependencies
- https://github.com/nervosnetwork/godwoken/pull/808
- https://github.com/nervosnetwork/godwoken-scripts/pull/134

## Tests

test code:
https://github.com/Flouse/godwoken-polyjuice/blob/47865fc349d9485021b3e75205cfd621c1ada8e1/polyjuice-tests/src/test_cases/pre_compiled_contracts.rs#L163-L697

cmd:
```bash
cd polyjuice-tests/
cargo test --release -- test_bn256_add --nocapture 2>&1
cargo test --release -- test_bn256_scalar_mul --nocapture 2>&1
cargo test --release -- test_bn256_pairing --nocapture 2>&1
```

| Test Case                      | Elapsed Time (ms) | Used Gas | Used Cycles (virtual+execution) |
| ------------------------------ | ----------------: | -------: | ------------------------------: |
| Bn256Add                       |                 1 |     ~27K |                           ~693K |
| Bn256ScalarMul                 |                <2 |     ~31K |                           ~690K |
| Bn256Pairing one_point         |                 5 |    ~105K |                           ~918K |
| Bn256Pairing two_point_match_2 |                 6 |    ~141K |                          ~1024K |
| Bn256Pairing two_point_match_3 |                 7 |    ~141K |                          ~1024K |
| Bn256Pairing two_point_match_4 |                 8 |    ~141K |                          ~1024K |
| Bn256Pairing ten_point_match_1 |                22 |    ~432K |                          ~1874K |
| Bn256Pairing ten_point_match_2 |                23 |    ~436K |                          ~1874K |
| Bn256Pairing ten_point_match_3 |                 8 |    ~142K |                          ~1024K |

### CODE REVIEW

File: polyjuice-tests/src/helper.rs [hunk 1]
```diff
@@ -359,7 +359,12 @@ impl PolyjuiceArgsBuilder {
 }
 
 pub fn setup() -> (Store, DummyState, Generator, u32) {
-    let _ = env_logger::try_init();
+    // If you want to watch the [contract debug] logs in Polyjuice,
+    // please change the log level from `info` to `debug`.
+    // then run `cargo test -- [test_filter] --nocapture`, 
+    // or run `RUST_LOG=debug cargo test -- [test_filter] --nocapture` directly
```
Comment (TheWaWaR) [inline polyjuice-tests/src/helper.rs:365]: better with `RUST_LOG=gw=debug`, the `gw` prefix can filter out a lot unrelated log.

**Review by TheWaWaR — DISMISSED** [review]

File: c/contracts.h [hunk 2]
```diff
@@ -5,12 +5,19 @@
...
 
 #include "polyjuice_utils.h"
 #include "sudt_contracts.h"
 #include "other_contracts.h"
 
+/**
+ * alt_bn128_staticlib
+ * @see https://github.com/mohanson/bn/blob/d1c73cc198a14ca894a37ba99df88a9fb812a43c/alt_bn128_staticlib/src/ethereum.rs
+ */
+extern "C" uint32_t alt_bn128_add(const uint8_t *data, uint32_t data_len, uint8_t *output);
```
Comment (XuJiandong) [inline c/contracts.h:17]: It's better to return "uint64_t" or "int" as the error code.

File: c/generator.c [hunk 3]
```diff
@@ -13,4 +13,12 @@
 
 #include "polyjuice.h"
 
-int main() { return run_polyjuice(); }
+int main() {
+  // A temporal patch to solve https://github.com/nervosnetwork/ckb-vm/issues/97
+  CKB_SP_ALIGN;
```
Comment (XuJiandong) [inline c/generator.c:18]: This align can be removed, I think.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: c/contracts.h (hunk 2 above)
Comment (Flouse) [reply c/contracts.h:17]: The extern "C" functions use `u32` as a return code:
https://github.com/mohanson/bn/blob/master/alt_bn128_staticlib/src/ethereum.rs#L2

should I use int in c/contracts.h?

Comment (XuJiandong) [reply c/contracts.h]: it seems `u32` is OK here.

**Review by jjyr — DISMISSED** [review]

**Review by magicalne — DISMISSED** [review]

**Review by zeroqn — DISMISSED** [review]

**Comment by jjyr** [comment]
Please rebase

**Review by jjyr — APPROVED** [review]

**Review by XuJiandong — APPROVED** [review]

**Review by magicalne — APPROVED** [review]

**Pull request merged**